urbanairship-react-native/js/AttributeEditor.ts

```diff
@@ -0,0 +1,100 @@
+/* Copyright Airship and Contributors */
+
+'use strict';
+
+/**
+ * Attribute operation
+ * @hidden
+ */
+export interface AttributeOperation {
+  /**
+   * The operation name.
+   */
+  action: string;
+  /**
+   * The attribute key.
+   */
+  key: string;
+  /**
+   * The attribute value, if avaialble.
+   */
+  value?: string | number | Date;
+  /**
+   * The attribute type, if available.
+   */
+  type?: "string" | "number" | "date";
+}
+
+/**
+ * Editor for attributes.
+ */
+export class AttributeEditor {
+  onApply: (operations: AttributeOperation[]) => void;
+  operations: AttributeOperation[];
+
+  /**
+   * AttributeEditor constructor
+   *
+   * @hidden
+   * @param onApply The apply function
+   */
+  constructor(onApply: (operations: AttributeOperation[]) => void) {
+    this.onApply = onApply;
+    this.operations = [];
+  }
+
+  /**
+   * Adds an attribute.
+   *
+   * @param value The attribute value.
+   * @param name The attribute name.
+   * @return The attribute editor instance.
+   */
+  setAttribute(name: string, value: string | number | boolean | Date): AttributeEditor {
+    var attributeValue: string | number | Date;
+    var attributeType: "string" | "number" | "date";
+
+    if (typeof value != "boolean") {
+      attributeValue = value;
+      if (typeof value === "string") {
+        attributeType = "string";
+      } else if (typeof attributeValue === "number") {
+        attributeType = "number";
+      } else if (value instanceof Date) {
+        // JavaScript's date type doesn't pass through the JS to native bridge.
+        // Dates are instead serialized as milliseconds since epoch.
+        attributeType = "date";
+        attributeValue = value.getTime();
+      } else {
+        throw "Unsupported attribute type: " + typeof attributeValue;
+      }
+    } else {
+      // No boolean attribute type. Convert value to string.
+      attributeValue = value.toString();
+      attributeType = "string";
+    }
```

Comment:
I think this if would read better if all of the tests were positive, i.e.
```suggestion
    if (typeof value === "string") {
      attributeType = "string";
      attributeValue = value;
    } else if (typeof attributeValue === "number") {
      attributeType = "number";
      attributeValue = value;
    } else if (typeof value === "boolean") {
      // No boolean attribute type. Convert value to string.
      attributeType = "string";
      attributeValue = value.toString();
    } else if (value instanceof Date) {
      // JavaScript's date type doesn't pass through the JS to native bridge.
      // Dates are instead serialized as milliseconds since epoch.
      attributeType = "date";
      attributeValue = value.getTime();
    } else {
      throw "Unsupported attribute type: " + typeof attributeValue;
    }
```

Reply:
Your example actually won't compile because attributeValue hasn't been assigned to in the final else clause. That's why boolean is special cased (underlyingly, it's not a valid attribute type). I can rewrite it to use a positive test for boolean-ness though.